File: aavt/src/main/java/com/wuwang/aavt/core/Observable.java
package com.wuwang.aavt.core;

import java.util.ArrayList;
import java.util.Observer;

/*
 * Created by Wuwang on 2017/10/23
 */
public class Observable<Type> implements IObservable<Type> {

    private ArrayList<IObserver<Type>> temp;

    @Override
    public void addObserver(IObserver<Type> observer) {
        if(temp==null){
            temp=new ArrayList<>();
        }
        temp.add(observer);
    }

    @Override
    public boolean removeObserver(IObserver<Type> observer) {
        if(temp==null){
            return true ;
        }
        return temp.remove(observer);
    }

    @Override
    public int clearObserver() {
        if(temp == null) return 0 ;
        int size = temp.size() ;
        temp.clear();
        temp = null ;
        return size;
    }

    @Override
    public void notify(Type type) {
        for (IObserver<Type> t:temp){
            t.onCall(type);
        }
    }

}

File: common/src/main/java/com/wyz/common/core/Observable.kt
package com.wyz.common.core

import com.wyz.common.api.IObservable
import com.wyz.common.api.IObserver
import java.util.*

open class Observable<T> : IObservable<T> {

    private val observers: ArrayList<IObserver<T>>

    constructor() {
        this.observers = arrayListOf()
    }

    override fun addObserver(observer: IObserver<T>) {
        synchronized(observers){
            observers.add(observer)
        }
    }

    override fun removeObserver(observer: IObserver<T>): Boolean {
        synchronized(observers){
            return observers.remove(observer)
        }
    }

    override fun clearObserver() {
        synchronized(observers){
            observers.clear()
        }
    }

    override fun notify(type: T) {
        synchronized(observers){
            observers.forEach { it.run(type) }
        }
    }
}